File: lib/src/main/java/com/orientsec/easysocket/session/QueuedWriter.java
package com.orientsec.easysocket.session;

import androidx.annotation.NonNull;

import com.orientsec.easysocket.EasyExecutor;
import com.orientsec.easysocket.client.BaseSocketClient;
import com.orientsec.easysocket.error.EasyException;
import com.orientsec.easysocket.error.ErrorCode;
import com.orientsec.easysocket.error.ErrorType;
import com.orientsec.easysocket.task.OperableTask;

import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.concurrent.Executor;

/**
 * A writer implementation that manages a queue of tasks to be written to a socket.
 * This class ensures that tasks are written sequentially and handles errors during the writing
 * process.
 */
class QueuedWriter implements Writer {
    // The session associated with this writer
    private final OperableSession session;
    // The socket used for writing data
    private final Socket socket;
    // Executor for handling write operations
    private final Executor writeExecutor;
    // Runner for scheduling tasks on the main thread
    private final EasyExecutor mainExecutor;
    // Queue for storing tasks to be written
    private final Deque<OperableTask<?>> writingQueue = new ArrayDeque<>();
    // Flag indicating whether a write operation is currently in progress
    private boolean isWriting = false;

    /**
     * Constructs a QueuedWriter instance with the specified parameters.
     *
     * @param session The session associated with this writer.
     * @param socket  The socket used for writing data.
     * @param client  The BaseSocketClient instance providing options and executors.
     */
    public QueuedWriter(OperableSession session, Socket socket, BaseSocketClient client) {
        this.session = session;
        this.socket = socket;
        this.writeExecutor = client.getOptions().getWriteExecutor();
        this.mainExecutor = client.getMainExecutor();
    }

    /**
     * Submits a task to the writing queue.
     *
     * @param task The task to be written.
     */
    @Override
    public void submit(@NonNull OperableTask<?> task) {
        enqueue(task);
    }

    /**
     * Adds a task to the writing queue and schedules the next write operation if none is in
     * progress.
     *
     * @param task The task to be added to the queue.
     */
    private void enqueue(@NonNull OperableTask<?> task) {
        writingQueue.add(task);
        if (!isWriting) {
            isWriting = true;
            scheduleNextWrite();
        }
    }

    /**
     * Schedules the next write operation from the queue.
     * If the queue is empty, marks the writer as not writing.
     */
    private void scheduleNextWrite() {
        if (writingQueue.isEmpty()) {
            isWriting = false;
            return;
        }

        OperableTask<?> nextTask = writingQueue.removeFirst();
        writeExecutor.execute(() -> write(nextTask));
    }

    /**
     * Writes data to the socket in the write thread.
     *
     * @param task The task containing the data to be written.
     */
    private void write(OperableTask<?> task) {
        mainExecutor.execute(task::onSendStart);
        try {
            OutputStream outputStream = socket.getOutputStream();
            outputStream.write(task.getData());
            outputStream.flush();
            mainExecutor.execute(task::onSendSuccess);
        } catch (IOException e) {
            session.getLogger().e("socket write error", e);
            EasyException ex = EasyException.create(ErrorCode.WRITE_ERROR, ErrorType.CONNECT,
                    "socket write aborted", session.getSuffix(), e);
            mainExecutor.execute(() -> {
                task.onSendFailure(ex);
                session.close(ex);
            });
        } finally {
            // Schedules the next write operation on the main thread
            mainExecutor.execute(this::scheduleNextWrite);
        }
    }

    /**
     * Cancels all tasks in the writing queue.
     */
    public void cancelAll() {
        writingQueue.clear();
    }

    /**
     * Cancels a specific task in the writing queue.
     *
     * @param task The task to be canceled.
     */
    @Override
    public void cancel(@NonNull OperableTask<?> task) {
        boolean removeFromWritingQueue = writingQueue.remove(task);
        session.getLogger().i("removed task " + task.getTaskId() +
                " from writing queue:" + removeFromWritingQueue);
    }
}

File: lib/src/main/kotlin/com/orientsec/easysocket/socket/QueuedWriter.kt
package com.orientsec.easysocket.socket

import com.orientsec.easysocket.error.EasyException
import com.orientsec.easysocket.error.ErrorCode
import com.orientsec.easysocket.error.ErrorType
import com.orientsec.easysocket.session.OperableSession
import com.orientsec.easysocket.session.Writer
import com.orientsec.easysocket.task.OperableTask
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException
import java.io.OutputStream
import java.net.Socket
import java.util.ArrayDeque
import java.util.Deque

/**
 * A writer implementation that manages a queue of tasks.
 * This class ensures sequential writing by scheduling the next task only after
 * the current one completes, mirroring the original Java implementation.
 */
class QueuedWriter(
    private val session: OperableSession,
    private val socket: Socket,
    private val scope: CoroutineScope
) : Writer {
    // 使用 Deque 作为任务队列
    private val writingQueue: Deque<OperableTask<*>> = ArrayDeque()

    // 写入状态标识
    private var isWriting = false

    /**
     * 清空队列并重置状态。
     */
    fun shutdown() {
        writingQueue.clear()
        isWriting = false
    }

    override fun submit(task: OperableTask<*>) {
        enqueue(task)
    }

    /**
     * 入队逻辑：如果正在写入则加入队列，否则立即开始发送。
     */
    private fun enqueue(task: OperableTask<*>) {
        if (isWriting) {
            writingQueue.add(task)
        } else {
            isWriting = true
            task.onSendStart()
            write(task)
        }
    }

    /**
     * 调度下一个任务：从队列头部取出任务并执行。
     */
    private fun scheduleNextWrite() {
        val nextTask = writingQueue.pollFirst()
        if (nextTask == null) {
            isWriting = false
        } else {
            nextTask.onSendStart()
            write(nextTask)
        }
    }

    /**
     * 执行实际的写入操作。
     */
    private fun write(task: OperableTask<*>) {
        scope.launch {
            try {
                // 在 IO 线程池中执行阻塞写入
                withContext(Dispatchers.IO) {
                    val outputStream: OutputStream = socket.getOutputStream()
                    val data = task.data ?: throw IOException("Task data is null")
                    outputStream.write(data)
                    outputStream.flush()
                }
            } catch (e: IOException) {
                // 写入失败：执行失败处理并关闭 Session
                session.logger.w("socket write error ", e)
                val ex = EasyException.Companion.invoke(
                    ErrorCode.WRITE_ERROR, ErrorType.CONNECT,
                    "socket write aborted", session.suffix, e
                )
                task.onSendFailure(ex)
                session.close(ex)
            }
            // 写入成功：执行回调并调度下一个
            task.onSendSuccess()
            scheduleNextWrite()
        }
    }

    override fun cancel(task: OperableTask<*>) {
        val removed = writingQueue.remove(task)
        session.logger.i("removed task ${task.taskId} from writing queue: $removed")
    }
}